src/send/compression.rs

```diff
@@ -37,110 +13,83 @@ pub(crate) fn bytes2type(bytes: [u8; 5]) -> Format {
-            ))
-        }
-    } else {
-        pub(crate) fn new_gz_encoder<'a>(_: Box<dyn io::Write + Send + 'a>, _: Level) -> Result<Box<dyn io::Write + Send + 'a>, Error> {
-            Err(Error::FeatureDisabled)
-        }
-
-        pub(crate) fn new_gz_decoder<'a>(_: Box<dyn io::Read + Send + 'a>) -> Result<(Box<dyn io::Read + Send + 'a>, Format), Error> {
-            Err(Error::FeatureDisabled)
-        }
-    }
-}
-
-cfg_if! {
-    if #[cfg(feature = "bz2")] {
-        pub(crate) fn new_bz2_encoder<'a>(out: Box<dyn io::Write + Send + 'a>, level: Level) -> Result<Box<dyn io::Write + Send + 'a>, Error> {
-            Ok(Box::new(bzip2::write::BzEncoder::new(
-                out,
-                level.into(),
-            )))
-        }
-
-        pub(crate) fn new_bz2_decoder<'a>(
-            inp: Box<dyn io::Read + Send + 'a>,
-        ) -> Result<(Box<dyn io::Read + Send + 'a>, Format), Error> {
-            Ok((
-                Box::new(bzip2::read::BzDecoder::new(inp)),
-                Format::Bzip,
-            ))
-        }
-    } else {
-        pub(crate) fn new_bz2_encoder<'a>(_: Box<dyn io::Write + Send + 'a>, _: Level) -> Result<Box<dyn io::Write + Send + 'a>, Error> {
-            Err(Error::FeatureDisabled)
-        }
-
-        pub(crate) fn new_bz2_decoder<'a>(_: Box<dyn io::Read + Send + 'a>) -> Result<(Box<dyn io::Read + Send + 'a>, Format), Error> {
-            Err(Error::FeatureDisabled)
-        }
-    }
-}
-
-cfg_if! {
-    if #[cfg(feature = "lzma")] {
-    pub(crate) fn new_lzma_encoder<'a>(out: Box<dyn io::Write + Send + 'a>, level: Level) -> Result<Box<dyn io::Write + Send + 'a>, Error> {
-            Ok(Box::new(liblzma::write::XzEncoder::new(out, level.into())))
+impl_format!(
+    gz,
+    "gz",
+    crate::send::compression::Format::Gzip,
+    flate2::write::GzEncoder::new,
+    flate2::read::MultiGzDecoder::new,
+    std::io::Read | Send,
+    std::io::Write | Send,
+    crate::send::compression::Format
+);
+
+impl_format!(
+    bz2,
+    "bz2",
+    crate::send::compression::Format::Bzip,
+    bzip2::write::BzEncoder::new,
+    bzip2::read::MultiBzDecoder::new,
+    std::io::Read | Send,
+    std::io::Write | Send,
+    crate::send::compression::Format
+);
+
+impl_format!(
+    lzma,
+    "lzma",
+    crate::send::compression::Format::Lzma,
+    liblzma::write::XzEncoder::new,
+    liblzma::read::XzDecoder::new,
+    std::io::Read | Send,
+    std::io::Write | Send,
+    crate::send::compression::Format
+);
+
+pub mod zstd {
+    /* standard use */
+    use std::io;
+
+    /* project use */
+    use crate::error::Error;
+
+    /* project use */
+    use crate::level::Level;
+
+    use super::Format;
+
+    #[cfg(feature = "zstd")]
+    pub(crate) fn encoder<'a>(
+        out: Box<dyn io::Write + Send + 'a>,
+        level: Level,
+    ) -> Result<Box<dyn io::Write + Send + 'a>, Error> {
+        Ok(Box::new(
+            zstd::stream::write::Encoder::new(out, level.into())?.auto_finish(),
```

Comment:
Couldn't do `zstd` because it has the extra `.auto_finish()`, need to look more into it